pkg/controller/job/job_controller.go

```diff
@@ -985,10 +985,16 @@ func (jm *Controller) newFailureCondition(reason, message string) *batch.JobCond
 
 func (jm *Controller) newSuccessCondition() *batch.JobCondition {
 	cType := batch.JobComplete
+	var reason, message string
 	if delayTerminalCondition() {
 		cType = batch.JobSuccessCriteriaMet
+
+		// To avoid the breaking the existing Completion condition, the CompletionsReached reason and message are set
+		// only when the condition type is SuccessCriteriaMet.
+		reason = batch.JobReasonCompletionsReached
```

Comment:
No strong view here, but I think delayTerminalCondition is rather orthogonal, and I would see the new reason under the SuccessPolicy feature, because it introduces the new "reasons".

Reply:
@mimowo Your suggestion meant the like this?

```golang
		if delayTerminalCondition() {
			cType = batch.JobSuccessCriteriaMet
		}
		if cType == batch.SuccessCriteriaMet {
			reason = batch.JobReasonCompletionsReached
			message = "Reached to completions"
		}
```

Reply:
I'm thinking about 
```
		if delayTerminalCondition() {
			cType = batch.JobSuccessCriteriaMet
		}
		if feature.DefaultFeatureGate.Enabled(features.JobSuccessPolicy) {
			reason = batch.JobReasonCompletionsReached
			message = "Reached to completions"
		}
```

Reply:
That looks good to me. But I feel weird that only reason and message are guarded by SuccessPolicy feature gate.
Because the SuccessCriteriaMet is added regardless of SuccessPolicy feature gate.

Is there any discussion of the reason why the SuccessCriteriaMet is not guarded by the SucceessPolicy feature gate?

Reply:
We decided to extend the use of `SuccessCriteriaMet` and `FailureTarget` conditions for all failure scenarios. They are now used to communicate the Job is going to be failed or succeeded as soon as Job controller knows. This is part of fixing https://github.com/kubernetes/kubernetes/issues/123775. The approach was discussed on the SIG-[apps meeting](https://docs.google.com/document/d/1LZLBGW2wRDwAfdBNHJjFfk9CFoyZPcIYGWU7R1PQ3ng/edit) May 13, see also the [slides](https://docs.google.com/presentation/d/1VsHdByAo5O0lI4ZdTox2T5-5xW1npNE-TpFU13j5NQg/edit?resourcekey=0-9E3pfK8RngCpqdctOmgYvQ#slide=id.g2dbc89df2f5_0_25).

It is also part of the KEP for ManagedBy & PodReplacementPolicy (see [here](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-for-batch-jobs#terminating-pods-and-terminal-job-conditions))

Reply:
Thank you for clarifying the reason! That makes sense.
I agree with introducing your proposal:

> ```golang
> 		if delayTerminalCondition() {
> 			cType = batch.JobSuccessCriteriaMet
> 		}
> 		if feature.DefaultFeatureGate.Enabled(features.JobSuccessPolicy) {
> 			reason = batch.JobReasonCompletionsReached
> 			message = "Reached to completions"
> 		}
> ```

Reply:
BTW, I think the `message` does not need to be under the feature gate, cause messages aren't meant to be interpreted by machines. Also, no strong view here, but I think this would read better `Reached expected number of succeeded pods`.

Reply:
> BTW, I think the message does not need to be under the feature gate, cause messages aren't meant to be interpreted by machines.

I'm still wondering if we should add a feature gate guard regardless of whether the field is machine readable.
The feature gate would be useful in the case where we find the any critical issues related to this feature.
In this situation, we can easily disable the enhanced message by default.

@soltysh @alculquicondor @atiratree  Anyway, let me know what you think.

>  Also, no strong view here, but I think this would read better `Reached expected number of succeeded pods`.

That sounds good to me.

Reply:
I agree with Michal here, the new reason (`JobReasonCompletionsReached`) is strictly tied with this particular feature (`JobSuccessPolicy`) so it should only apply when that feature is enabled.

Reply:
@soltysh Sorry for your confusion, here we are wondering if we should select which pattern.

A: All new fields are guarded by the feature gates.

```golang
		if delayTerminalCondition() {
			cType = batch.JobSuccessCriteriaMet
		}
		if feature.DefaultFeatureGate.Enabled(features.JobSuccessPolicy) {
			reason = batch.JobReasonCompletionsReached
			message = "Reached to completions"
		}
```

B: Only type and reason are guarded by the feature gates because the `message` is not expected to be machine-readable.

```golang
		if delayTerminalCondition() {
			cType = batch.JobSuccessCriteriaMet
		}
		if feature.DefaultFeatureGate.Enabled(features.JobSuccessPolicy) {
			reason = batch.JobReasonCompletionsReached
		}
		message = "Reached to completions"
```

Reply:
IMO coupling them together in option A seems better to limit the number of api changes between versions/features.

Reply:
> IMO coupling them together in option A seems better to limit the number of api changes between versions/features.

Thank you for checking this comment. I will apply the changes soon.